lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2297,6 +2299,212 @@ class ArchiveTool : public Tool {
   }
 };
 
+#pragma mark - StaleFileRemovalTool implementation
+
+class StaleFileRemovalCommand : public Command {
+  std::string description;
+
+  std::vector<std::string> expectedOutputs;
+  std::vector<std::string> filesToDelete;
+  bool computedFilesToDelete = false;
+
+  BuildValue priorValue;
+  bool hasPriorResult = false;
+
+  /// The cached signature, once computed -- 0 is used as a sentinel value.
+  std::atomic<uint64_t> cachedSignature{ 0 };
+
+  virtual uint64_t getSignature() {
+    uint64_t signature = cachedSignature;
+    if (signature != 0)
+      return signature;
+
+    // FIXME: Use a more appropriate hashing infrastructure.
+    using llvm::hash_combine;
+    llvm::hash_code code;
```

Comment:
This needs to initialize to `ExternalCommand::getSignature();`, otherwise it won't pick up the signature of shared things like the command output list.